Include params set in provide-client-param event handlers in dynamic context params for endpoint resolution

jonemo opened the pull request:
This PR moves the `provide-client-params.*` and `before-parameter-build.*` from immediately after endpoint resolution to immediately before endpoint resolution.

This addresses the fact that for endpoint rulesets that accept dynamic context parameters (currently `s3`, `s3-control`, `kinesis`, and `events`) any parameters changed or updated in an event handler for the `provide-client-params.*` were not accounted for in endpoint resolution because the event was only emitted _after_ endpoint resolution.

dlm6693 commented on botocore/client.py:
```diff
@@ -819,7 +819,6 @@ def _resolve_signature_version(self, service_name, resolved):
 
 
 class BaseClient:
-
```
Was this intentional?

jonemo replied:
I didn't add it intentionally. But when I saw and understood it, I intentionally did not revert it.

This is a change applied by the black formatter. black makes this change only in versions [23.1.0](https://black.readthedocs.io/en/stable/change_log.html#id8) and higher. 23.1.0 is the version that adheres to black's [2023 stable style](https://github.com/psf/black/issues/3407). I suspect but did not validate that [this change to empty line handling](https://github.com/psf/black/pull/3302) which is included in the 2023 stable style causes this change to be applied. botocore specifies the black version used for CI in [the precommit config file](https://github.com/boto/botocore/blob/a91a7bf/.pre-commit-config.yaml#L37). 

I decided to leave the change in for two reasons:
1. More and more contributors will update their local black version to >=23.1.0 and this change will continue sneaking into diffs (and then require unnecessary attention).
2. Versions of black <23.1.0 do not revert this change, so there is downside to keeping it.

dlm6693 replied:
Ack. If black did it and you reversed the change, it probably would've happened the next time this file was updated. Thanks for the explanation!

dlm6693 commented on botocore/client.py:
```diff
@@ -913,6 +912,11 @@ def _make_api_call(self, operation_name, api_params):
             'has_streaming_input': operation_model.has_streaming_input,
             'auth_type': operation_model.auth_type,
         }
+        api_params = self._emit_api_params(
+            api_params=api_params,
+            operation_model=operation_model,
+            context=request_context,
+        )
```
Maybe this could be called from within `_resolve_endpoint_ruleset` at the very start of it? I'm thinking in purely organizational terms just because `_make_api_call` is a rather long, complex function already.

dlm6693 reviewed (approved):
Just a couple of stylistic comments, but otherwise LGTM!

dlm6693 commented on botocore/signers.py:
```diff
@@ -663,7 +663,11 @@ def generate_presigned_url(
         ignore_signing_region=(not bucket_is_arn),
     )
-
+    params = self._emit_api_params(
```
I think using my suggestion of putting this in `_resolve_endpoint_ruleset` would accomplish this no?

dlm6693 replied:
Or does this still need to happen immediately after endpoint resolution for presigned URLs?

jonemo commented:
@dlm6693 @nateprewitt Added two commits that consolidate the previously duplicated code. This comes at the cost of a change to the API of the `_convert_to_request_dict` method. The signature remains backwards compatible, but the `endpoint_url` argument is no longer used inside the function. There is a small chance that this will break anyone who relies on this "private" internal interface.

While making this change I also fixed the presign functions which were actually still affected by the original bug and added test coverage for this oversight.

jonemo commented:
After reviewing potential impact of the API change contained in bd44524, @nateprewitt and I decided to revert to the less well organized state that has been discussed above. Note that f8e01a6 is _not_ an exact line-for-line revert because it keeps the bugfix described in https://github.com/boto/botocore/pull/2920#issuecomment-1539183503.

dlm6693 commented on botocore/signers.py:
```diff
@@ -656,14 +656,18 @@ def generate_presigned_url(
         context,
         ignore_signing_region=(not bucket_is_arn),
     )
-
```
Was this also black?

jonemo replied:
No. Brought the two empty lines back in d8ed89d.

dlm6693 commented:
@jonemo still LGTM after the latest commit 👍

nateprewitt reviewed (approved):
:shipit:
